Create nullable migration tutorial (#10383)
dotnet/docs: docs/csharp/tutorials/nullable-reference-types.md

## Prerequisites

> You’ll need to set up your machine to run .NET Core, including the C# 8.0 beta compiler. The C# 8 beta compiler is available with [Visual Studio 2019 preview 1 and newer](https://visualstudio.microsoft.com/vs/preview/), or [.NET Core 3.0 preview 1](https://dotnet.microsoft.com/download/dotnet-core/3.0).

Reviewer: Nit: should the curly apostrophe in the first word be an ASCII one? (It is in the next paragraph.)

> You’ll need to set up your machine to run .NET Core, including the C# 8.0 beta compiler. The C# 8 beta compiler is available with [Visual Studio 2019 preview 1 and newer](https://visualstudio.microsoft.com/vs/preview/), or [.NET Core 3.0 preview 1](https://dotnet.microsoft.com/download/dotnet-core/3.0).

Reviewer 2: ```suggestion
You'll need to set up your machine to run .NET Core, including the C# 8.0 beta compiler. The C# 8 beta compiler is available with [Visual Studio 2019 preview 1 and newer](https://visualstudio.microsoft.com/vs/preview/), or [.NET Core 3.0 preview 1](https://dotnet.microsoft.com/download/dotnet-core/3.0).
```

## Create the application and enable nullable reference types

> Create a new console application either in Visual Studio or from the command line using `dotnet new console`. Name the application `NullableIntroduction`. Once you've created the application, you'll need to enable C# 8 beta features. Open the `csproj` file, and add a `LangVersion` element to the `PropertyGroup` element. You must opt into the **nullable reference types** feature, even in C# 8 projects. That's because once the feature is turned on, existing reference variable declarations become **non-nullable reference types**. While that decision will help find issues where existing code may not have proper null-checks, it may not accurately reflect your original design intent. You turn on the feature by setting the `NullableContextOptions` element to `enable`:

Reviewer 3: just removing the comma from "Open the file and add"...

```suggestion
Create a new console application either in Visual Studio or from the command line using `dotnet new console`. Name the application `NullableIntroduction`. Once you've created the application, you'll need to enable C# 8 beta features. Open the `csproj` file and add a `LangVersion` element to the `PropertyGroup` element. You must opt into the **nullable reference types** feature, even in C# 8 projects. That's because once the feature is turned on, existing reference variable declarations become **non-nullable reference types**. While that decision will help find issues where existing code may not have proper null-checks, it may not accurately reflect your original design intent. You turn on the feature by setting the `NullableContextOptions` element to `enable`:
```

```xml
<LangVersion>8.0</LangVersion>

> <NullableContextOptions>enable</NullableContextOptions>

Reviewer: Earlier you've mentioned VS 2019 preview 1 - this would require VS 2019 preview 2. It might be simplest just to require preview 2 rather than handle both.

Author: Good catch. I'll continue tracking only the latest preview.

[!code-csharp[AddQuestions](../../../samples/csharp/NullableIntroduction/NullableIntroduction/Program.cs#AddQuestions)]

> Because the entire project is a nullable enabled context, you'll get warnings when you pass `null` to any method expecting a non-nullable reference type. Try it by adding the following line to `Main`:

Reviewer: "the entire project is a nullable enabled context" => "the entire project is in a nullable enabled context"? Or "the nullable context is enabled for the entire project"?
(It's possible that what you've written is exactly what you intended - contexts can be like that, I guess. It just reads a little oddly to me.)

The storage for the survey answers is a `Dictionary<int, string>?`, indicating that it may be null. You're using the new language feature to declare your design intent, both to the compiler and to anyone reading your code later. If you ever dereference `surveyResponses` without checking for the null value first, you'll get a compiler warning. You don't get a warning in the `AnswerSurvey` method because the compiler can determine the `surveyResponses` variable was set to a non-null value above.

> Using null for missing answers highlights a key point for working with nullable reference types: your goal is not to remove all `null` values from your program. Rather, your goal is to ensure that the code you write expresses the intent of your design. Missing values are a necessary concept to express in your code. The `null` value is a clear way to express those missing values. Trying to remove all `null` values only leads to defining some other way to express those missing values without null.

Reviewer 2: ```suggestion
Using `null` for missing answers highlights a key point for working with nullable reference types: your goal isn't to remove all `null` values from your program. Rather, your goal is to ensure that the code you write expresses the intent of your design. Missing values are a necessary concept to express in your code. The `null` value is a clear way to express those missing values. Trying to remove all `null` values only leads to defining some other way to express those missing values without `null`.
```

## Next steps

> Learn more by mirating an existing application to use nullable reference types..

Reviewer: Nit: mirating => migrating

> Learn more by migrating an existing application to use nullable reference types..

Reviewer 3: should this use a colon instead?

```suggestion
Learn more by migrating an existing application to use nullable reference types:
```
